### PR TITLE
Solver options

### PR DESCRIPTION
**Pull request by meyer-nils**
This PR extends torch-fem’s differentiable solve stack by introducing implicit-adjoint autograd support for nonlinear Newton solves and by refining the solver APIs/docs/examples to use explicit differentiable_parameters inputs.

- Added an adjoint Newton-Raphson autograd operator for nonlinear solves via `newton_solve(...)`.
- Added gradient regression tests in `tests/test_gradients.py` for:
	- consistency between single-step and incremental gradients in mechanics,
	- finite and stable gradients for planar heat topology-style parameters.
- Refactored mechanics and heat integration interfaces in `base.py` to use explicit previous-step inputs and return updated integration-point fields instead of mutating global history tensors in-place.
- Renamed the autograd-enabled sparse linear solve entry point from `sparse_solve(...)` to `differentiable_sparse_solve(...)`, while `sparse_solve(...)` now denotes the backend sparse solve routine used by both forward and adjoint paths.
- Updated `solve(...)` and `time_integration(...)` to accept `differentiable_parameters` as either a single tensor or an iterable of tensors.
- Updated nonlinear and transient solve paths to use implicit adjoint logic with cleaner graph handling and optional cached sparse warm starts.
- Improved API and type annotations in solver internals (`sparse.py`, `base.py`) and expanded solver docstrings.
- Expanded differentiability documentation (`docs/differentiability.md`) with explicit sections on:
	- adjoint sparse linear solve,
	- adjoint Newton-Raphson for nonlinear FEM.
- Updated usage examples in `README.md` and `docs/getting_started.md` to pass `differentiable_parameters=...` in differentiable solve calls.
- Updated many notebooks and benchmark scripts/examples to match the current differentiable solve API (single tensor for single-parameter cases, tuple only for multi-parameter cases).

### CODE REVIEW

File: tests/test_gradients.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import torch
+
+from torchfem import Planar, PlanarHeat
+from torchfem.materials import IsotropicConductivity2D, IsotropicElasticityPlaneStress
+from torchfem.mesh import rect_quad
+
+torch.set_default_dtype(torch.float64)
+
```
[the remark this answers is not in the document]
Comment (meyer-nils) [reply tests/test_gradients.py:8]: @copilot apply changes based on [this feedback](https://github.com/meyer-nils/torch-fem/pull/34#discussion_r2956521880)
